pkg/statistics/histogram.go

```diff
@@ -1052,38 +1052,41 @@ func (hg *Histogram) OutOfRangeRowCount(
 		rightPercent = (math.Pow(boundR-actualL, 2) - math.Pow(boundR-actualR, 2)) / math.Pow(histWidth, 2)
 	}
 
-	totalPercent := leftPercent*0.5 + rightPercent*0.5
-	if totalPercent > 1 {
-		totalPercent = 1
-	}
-	rowCount = totalPercent * hg.NotNullCount()
+	totalPercent := min(leftPercent*0.5+rightPercent*0.5, 1.0)
 
-	// Upper bound logic
+	// Upper & lower bound logic.
+	upperBound := rowCount
+	if histNDV > 0 {
+		upperBound = hg.NotNullCount() / float64(histNDV)
+	}
 
 	allowUseModifyCount := sctx.GetSessionVars().GetOptObjective() != variable.OptObjectiveDeterminate
-	// Use the modifyCount as the upper bound. Note that modifyCount contains insert, delete and update. So this is
-	// a rather loose upper bound.
-	// There are some scenarios where we need to handle out-of-range estimation after both insert and delete happen.
-	// But we don't know how many increases are in the modifyCount. So we have to use this loose bound to ensure it
-	// can produce a reasonable results in this scenario.
-	if rowCount > float64(modifyCount) && allowUseModifyCount {
-		return float64(modifyCount)
-	}
-
-	// In OptObjectiveDeterminate mode, we can't rely on the modify count anymore.
-	// An upper bound is necessary to make the estimation make sense for predicates with bound on only one end, like a > 1.
-	// But it's impossible to have a reliable upper bound in all cases.
-	// We use 1/NDV here (only the Histogram part is considered) and it seems reasonable and good enough for now.
-	if !allowUseModifyCount {
-		var upperBound float64
-		if histNDV > 0 {
-			upperBound = hg.NotNullCount() / float64(histNDV)
-		}
-		if rowCount > upperBound {
-			return upperBound
-		}
+
+	if allowUseModifyCount && float64(modifyCount) > hg.NotNullCount() {
+		rowCount = totalPercent * float64(modifyCount)
+	} else {
+		rowCount = totalPercent * hg.NotNullCount()
 	}
-	return rowCount
+
+	if !allowUseModifyCount {
+		// In OptObjectiveDeterminate mode, we can't rely on the modify count anymore.
+		// An upper bound is necessary to make the estimation make sense for predicates with bound on only one end, like a > 1.
+		// We use 1/NDV here (only the Histogram part is considered) and it seems reasonable and good enough for now.
+		return min(rowCount, upperBound)
+	}
+	// If the modifyCount is large (as a percentage of the base table rows), then any out of range estimate is
+	// unreliable - because this indicates that our statistics are stale. These bounds are therefore approximations.
+	// ModifyCount counts all inserts/updates/deletes - this logic has a bias towards modifications being inserts.
+	// Assume a minimum of 1 value is found if we have a high modifyCount. This targest the situation where we search
+	// for a statisically out of range value, but inserts result in the value actually being in that modified range.
+	modifyPercent := min(float64(modifyCount)/hg.NotNullCount(), 1.0)
+	if modifyPercent == 1 || modifyPercent > totalPercent {
```

Comment:
The `modifyPercent > totalPercent` is a bit hard to understand for me.
`totalPercent` represents how large the out-of-range part is.
It looks like our criteria for stale stats are changing according to the size of the out-of-range part.

Reply:
As you say - totalPercent represents the out-of-range portion. But this is calculated based upon histograms that were collected at last ANALYZE. NotNullCount is the number of table rows minus NULLs - also collected from last ANALYZE.
realtimeRowCount and modifyCount are a count of the changes since last ANALYZE. It is my understanding that realtime/modifyCount do not update the histograms. Therefore, the larger the number of changes since the histograms were created - the greater the risk that any "totalPercent" (out of range estimate).
Therefore - what the above code is attempting to do - if the percentage of modifications to the original table (modifyCount/NotNullCount) is greater than the percentage we estimate out of range - it is possible that all of the rows we estimate to be out of range are actually within the modifyCount.

Reply:
Yes. Realtime row count and modify count don't update the histograms.


So, to my core question, if I understand it correctly, we are indeed using the relative size of modifications and out-of-range portion as criteria, which means that, if the modification is large, or if the out-of-range portion is small, we think the lower bound is needed.

Reply:
Correct - but I have attempted to address your concerns about the risk of using modifyCount in the latest revision.